### PR TITLE
perf(layers): extend CallSpec fast path to training=<bool> kwarg

### PR DESCRIPTION
**Pull request by pctablet505**
Closes the CallSpec-fast-path gap left open by #23198.

## Summary

#23198 added a `CallSpec` fast path that skips `inspect.Signature.bind` for
the common `layer(x)` call shape (one positional tensor, no kwargs). But
`fit()`/`evaluate()`/`predict()` (`torch/trainer.py`) call every layer with
`training=` passed **explicitly** — a shape the fast path's `not kwargs`
gate excludes. That's precisely the call shape that matters most in
practice: a 4-layer functional model measurably drops from 5 to 1
slow-path `CallSpec` binder invocations per forward pass once this PR's
gate admits it.

This widens the fast-path gate (`Layer.__call__` and `CallSpec._fast`) to
also accept the single well-known shape `kwargs == {"training": <exact
bool>}`, in addition to the existing no-kwargs case.

This PR is restacked onto the reworked #23198, which replaced
`inspect.Signature.bind` on the slow path with a pure-Python binder and
gave `CallSpec` a new `_call_param_specs` constructor argument — see
`## torch.compile impact` below for what that means for this PR.

## Mechanism

- `Layer.__call__`'s gate condition becomes `not kwargs or (len(kwargs) ==
  1 and (training is True or training is False))` — a flat boolean, no
  added nesting. The `len(kwargs) == 1` + identity checks together also
  guarantee the single key is `"training"` (any other key makes `.get`
  return `None`), so no separate key check is needed.
- `CallSpec._fast` gains a `training=None` keyword. When present, it builds
  `user_arguments_dict` in the same key order the slow path's
  `bound_args.arguments` (accepting layers) or
  `{**call_args, **bound_args.arguments}` (non-accepting layers) would
  produce, and overrides `arguments_dict["training"]` **after**
  construction (not inline) to preserve the same field-insertion order
  the slow path produces.
- Mirrors `CallSpec.__init__`'s kwargs-pop side effect: if the layer's
  `call()` doesn't accept `training`, it's popped from the caller's
  `kwargs` dict, exactly like the slow path — downstream code (context
  resolution, the eventual `self.call(**kwargs)`, Node construction) sees
  the same dict either path produces. This holds with #23198's reworked
  slow path too: the slow path's binder still produces the same
  `bound_args`-equivalent result this mirrors.
- Everything else — `training=None`, non-bool values (`np.bool_`, a
  tensor), a second kwarg, `**kwargs` signatures, multi-tensor calls,
  subclass-registered context args (`_register_call_context_args`) — falls
  through to the unchanged slow path unmodified.

## Why it is safe

- No priority-class change: the `training` value stays user-explicit
  (recorded in `user_arguments_dict`), so context resolution and
  `call_context.set_value` behave identically to the slow path — unlike a
  context-inheritance shortcut, nested-model and explicit-`None` edge
  cases are structurally unreachable here, not just tested around.
- Backend-agnostic by construction: the diff touches only
  `Layer.__call__`/`CallSpec._fast`, above any backend dispatch, with zero
  `backend.backend()` guards. Verified with a live `layer_test.py` run on
  torch, tensorflow, and jax — all clean.
- Subclass-bypass safe: none of the attributes the fast path reads
  (`_call_spec_fast_path`, `_call_has_context_arg`, `_call_arg_defaults`,
  `_call_arg_names`, `_call_first_arg_name`, `_call_context_args`) are
  `@property`-backed anywhere in `layer.py`; live-verified with a
  subclassed `Dense` overriding `call()` that the fast path correctly
  dispatches to the override.
- `functional.py` is untouched.
- The redundant `"training" in defaults` re-check inside `CallSpec._fast` is
  fixed to reuse the already-computed `accepts_training` flag instead of
  re-deriving it.

## torch.compile impact

This PR restacks onto the reworked #23198, whose slow-path `CallSpec`
binder took a new `_call_param_specs` constructor argument. This PR's own
`CallSpec._fast` widening doesn't construct `CallSpec` directly (it fills
fields in place, same as before), so it needed no change for that
signature — but the restack was verified for it explicitly rather than
assumed.

**Neutralization:** a break census comparing this PR's restacked tip
against the reworked #23198 tip shows a **break-count delta of 0** (CNN 9
→ 9 breaks, LLM 12 → 12 breaks, same break sites) — this PR adds no
`torch.compile` graph breaks on top of the reworked #23198. This is a
delta, not a zero-break claim: at this tip, neither #23198 alone nor this
PR reaches zero breaks by itself (see the compile timing below, where the
isolated state is still consistently slower under `torch.compile` than
eager) — the zero-break state is reached further up the chain, once
#23189 and #23201 are also present (that combination is what the stacked
numbers below reflect). (Prior to the #23198 rework, this PR's gate
widening was measured as adding one additional `LLM` graph break when
stacked on the earlier 13-PR combined branch; that is no longer the case
once it sits on top of the reworked, binder-based #23198.)

## Tests

`layer_test.py` — parity vs the slow path (field-by-field, both an
accepting and a non-accepting layer, both `training` values, eager and
symbolic inputs, including the kwargs-pop side effect), negative cases
(`training=None`, `np.bool_`, tensor `training`, an extra kwarg, and a
subclass-registered context arg all correctly stay on the slow path), a
numeric end-to-end parity test, and a deterministic bind-count assertion.
Ported onto the reworked #23198 base: the parity tests now construct
`CallSpec` with the `_call_param_specs` argument the reworked constructor
requires, rather than a bare `Signature`.

Full blast radius, torch backend: `layers/` 2017 passed / 224 skipped / 1
xpassed, `models/` 151 passed, `trainers/` (the actual `fit`/`evaluate`
path this targets) 360 passed / 45 skipped. Cross-backend sanity
(`layer_test.py` only): torch 92 passed / 2 skipped, jax 93 passed / 1
skipped, tensorflow 93 passed / 1 skipped, numpy 91 passed / 3 skipped.

## Benchmarks

The primary evidence here is the deterministic mechanism, not a GPU
wall-clock number — stated plainly, not smoothed over:

**Deterministic call-count proof** (reproducible, machine-independent):
`model(x, training=False)` on a 4-layer functional model — slow-path
`CallSpec` binder (`_bind_call_arguments`) invocations per forward: **5 →
1** with this fix, matching the bare `model(x)` case exactly. `model(x)`
itself is unaffected (1 → 1), confirming no regression to the
already-fast path. (This supersedes an earlier `inspect.Signature.bind`
count of 14 → 3: the reworked #23198 replaced `Signature.bind` on the slow
path with the pure-Python `_bind_call_arguments`, so `Signature.bind` is
no longer called on any path at this PR's current tip, and the binder
invocation count is the metric that still discriminates.)

**GPU wall-clock (8-pair interleaved, same methodology as the rest of the
#22561 series, NULL=`null-calibration-8pair-rtxpro6000-2026-07-19-run2.json`,
RTX PRO 6000 Blackwell, torch 2.13.0+cu130), measured at this PR's
pre-restack tip on 2026-07-19 (directionally valid after the restack: the
restack only made the slow path cheaper, it did not change what this PR's
own gate widening does):** isolated (this PR's branch, carrying #23198
beneath it, vs master) — CNN forward lands as **NOISE** (0.97x ratio,
doesn't clear the noise floor), LLM forward is **REAL** at 1.042x faster,
LLM generate-32 is **REAL** at 1.045x faster, all with 8/8 pair agreement.
Stacked (the full combined branch with this PR on top, vs master) is
**REAL** on all three workloads with 8/8 agreement: CNN forward 1.239x
faster, LLM forward 1.373x faster, LLM generate-32 1.376x faster. This
mirrors #23198 itself, which was justified by the same kind of
eager-dispatch-cost argument; this PR is the direct completion of that
argument, and at the stacked state the wall-clock evidence now clears the
noise floor across the board, not just directionally.

## torch.compile performance

PR #23288: master vs PR alone vs PR+parents, compiled, Keras torch GPU

Same-process eager-then-compile timing, isolated (this PR's branch, which
carries the reworked #23198 beneath it) and stacked (this PR on top of the
full combined branch, which also carries #23189 and #23201):

| state | CNN forward (eager / compile / speedup) | LLM forward (eager / compile / speedup) | LLM generate-32 (eager / compile / speedup) |
|---|---|---|---|
| master | 1.333 ms / 3.096 ms / 0.43x | 3.732 ms / 6.342 ms / 0.59x | 123.672 ms / 195.888 ms / 0.63x |
| isolated (this PR + #23198) | 1.128 ms / 2.293 ms / **0.49x** | 3.265 ms / 4.920 ms / **0.66x** | 107.742 ms / 157.872 ms / **0.68x** |
| stacked (this PR + #23189 + #23198 + #23201 + rest) | 0.892 ms / 0.372 ms / **2.40x** | 2.258 ms / 0.620 ms / **3.64x** | 78.448 ms / 10.052 ms / **7.80x** |

Stated plainly, not smoothed over: `torch.compile` still **hurts** at the
isolated state — every workload is 1.47–2.03x slower compiled than eager,
matching master's own eager-beats-compiled pattern, because the isolated
branch (this PR + the reworked #23198 alone) is not zero-break. Once the
full stack is applied — the same 9-CNN/12-LLM-break count this PR
contributes zero to, but with #23189 and #23201 also present — `compile`
flips to faster than eager everywhere: 0.42x of eager time for CNN
forward, 0.27x for LLM forward, and 0.13x (a 7.8x speedup) for LLM
generate-32. No cell stalled or errored in either run
(`compiled_error` is `null` throughout). This PR's own contribution to
that flip is exactly the zero break-count delta above — it does not
individually change whether compile wins or loses, it just doesn't cost
anything either way, and the win shows up once it's combined with the
rest of the chain.

## Series context

Found during a secondary root-cause investigation into #22561 after the
original 13-PR series was published — not one of the original 13, but
closes a real gap in #23198's own stated goal (the eager `CallSpec` fast
path that skips `signature.bind`). Depends on #23198's
`_call_spec_fast_path`/`CallSpec._fast` infrastructure; it stacks this
PR's own commit directly on top of #23198's branch (post-rework).

## Contributor Agreement

- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

### CODE REVIEW

File: keras/src/layers/layer.py [hunk 1]
```diff
@@ -332,6 +332,51 @@ def __init__(
...
+            # structure would land in tensor_arguments_dict in the slow path,
+            # so such signatures must fall through.
+            defaults_are_safe = all(
+                not is_backend_tensor_or_symbolic(v)
+                and not (
+                    tree.is_nested(v)
+                    and (not hasattr(v, "__len__") or len(v) > 0)
+                )
+                for v in self._call_arg_defaults.values()
+            )
```
[the remark this answers is not in the document]
Comment (pctablet505) [reply keras/src/layers/layer.py:376]: Thanks — this is a reasonable refinement, but I am keeping it out of this PR. The change here is scoped narrowly to admitting the `training=<bool>` kwarg into the existing fast-path gate. Widening `defaults_are_safe` to accept non-tensor nested defaults is an independent optimization: in practice almost no layer `call()` signatures carry nested defaults (they are overwhelmingly `None`/scalar, e.g. `training=None`, `mask=None`), so the added benefit is marginal, while the extra `tree.flatten`/`__len__` handling adds branching and new edge cases to a hot-path gate that is deliberately kept simple and provably conservative. Happy to revisit as a separate change if a concrete layer is shown to benefit; leaving the conservative check as-is here.

**Review by hertschuh — COMMENTED** [review]
This one should wait on and be rebased on https://github.com/keras-team/keras/pull/23198

**Comment by pctablet505** [comment]
This depended on #23198, so I think it should close.

Nothing here stands alone. Every symbol it touches — `_call_param_specs`, `_call_arg_defaults`, `CallSpec._fast` — exists only in #23198, and grepping master for them returns nothing. 520 of the 813 lines are #23198's; the rest only edit its code.

One correction to the description while I'm here: it says the binder goes from 5 calls to 1. Measured, it's 1 to 0.
